aten/src/ATen/native/LegacyDefinitions.cpp

```diff
@@ -702,4 +702,113 @@ Tensor remainder(const Tensor & self, const Tensor & other) {
+}
+
+Tensor max(const Tensor & self) {
+  return at::_th_max(self);
+}
+
+Tensor median(const Tensor & self) {
+  return at::_th_median(self);
+}
+
+std::tuple<Tensor &,Tensor &> sort_out(Tensor & values, Tensor & indices, const Tensor & self, int64_t dim, bool descending) {
+  return at::_th_sort_out(values, indices, self, dim, descending);
+}
+
+std::tuple<Tensor,Tensor> sort(const Tensor & self, int64_t dim, bool descending) {
+  return at::_th_sort(self, dim, descending);
+}
+std::tuple<Tensor &,Tensor &> topk_out(Tensor & values, Tensor & indices, const Tensor & self, int64_t k, int64_t dim, bool largest, bool sorted) {
+  return at::_th_topk_out(values, indices, self, k, dim, largest, sorted);
+}
+
+std::tuple<Tensor,Tensor> topk(const Tensor & self, int64_t k, int64_t dim, bool largest, bool sorted) {
+  return at::_th_topk(self, k, dim, largest, sorted);
+}
+
+Tensor all(const Tensor & self) {
+  return at::_th_all(self);
+}
+
+Tensor any(const Tensor & self) {
+  return at::_th_any(self);
+}
+
+Tensor & renorm_out(Tensor & result, const Tensor & self, Scalar p, int64_t dim, Scalar maxnorm) {
+  return at::_th_renorm_out(result, self, p, dim, maxnorm);
+}
+
+Tensor renorm(const Tensor & self, Scalar p, int64_t dim, Scalar maxnorm) {
+  return at::_th_renorm(self, p, dim, maxnorm);
+}
+
+Tensor unfold(const Tensor & self, int64_t dimension, int64_t size, int64_t step) {
+  return self._th_unfold(dimension, size, step);
+}
+
+bool equal(const Tensor & self, const Tensor & other) {
+  return at::_th_equal(self, other);
+}
+
+Tensor & pow_out(Tensor & result, const Tensor & self, const Tensor & exponent) {
+  return at::_th_pow_out(result, self, exponent);
+}
+
+Tensor pow(const Tensor & self, const Tensor & exponent) {
+  return at::_th_pow(self, exponent);
+}
+Tensor & pow_out(Tensor & result, Scalar self, const Tensor & exponent) {
+  return at::_th_pow_out(result, self, exponent);
+}
+
+Tensor pow(Scalar self, const Tensor & exponent) {
+  return at::_th_pow(self, exponent);
+}
+
+Tensor & normal_out(Tensor & output, const Tensor & mean, double std, Generator * generator) {
+  return at::_th_normal_out(output, mean, std, generator);
+}
+
+Tensor normal(const Tensor & mean, double std, Generator * generator) {
+  return at::_th_normal(mean, std, generator);
+}
+
+Tensor & normal_out(Tensor & output, double mean, const Tensor & std, Generator * generator) {
+  return at::_th_normal_out(output, mean, std, generator);
+}
+
+Tensor normal(double mean, const Tensor & std, Generator * generator) {
+  return at::_th_normal(mean, std, generator);
+}
+
+Tensor & normal_out(Tensor & output, const Tensor & mean, const Tensor & std, Generator * generator) {
+  return at::_th_normal_out(output, mean, std, generator);
+}
+
+Tensor normal(const Tensor & mean, const Tensor & std, Generator * generator) {
+  return at::_th_normal(mean, std, generator);
+}
+
+Tensor alias(const Tensor & self) {
+  return at::_th_alias(self);
```

Comment:
These calls now incur 2 type dispatches (once through the native function, and then another time through the _th_ version of the function. Ttype dispatch is ~10ns, so two are around ~20ns, but I've seen cases where doing a double type dispatch as much as 100ns).  This probably doesn't matter for most of the functions on this list, but something like alias is pretty fast. Is there a way for these to not do 2 dispatches?

Reply:
we had a mechanism to do this at one point ('is_type_dispatched'), which would pass 'this' (from Type) along to native function to avoid having to re-dispatch.  The issue is that adds Type as part of the schema when it's really an implementation detail.

I think a better approach is to just look through the list in LegacyDefinitions.cpp and figure out which ones actually matter in terms of having low overhead and reimplement those as _real_ native functions.

Reply:
I've actually been implementing the functions that matter by special casing them instead of writing native functions (see get_device, is_cuda, and is_sparse) but yeah that makes sense.